core/src/Streamly/Internal/Data/Stream/StreamD/Type.hs

```diff
@@ -1006,6 +1010,126 @@ iterateConcatMapDFS f stream = Stream step (stream, [])
                     (y:ys) -> return $ Skip (y, ys)
                     [] -> return Stop
 
+{-# ANN type IterateUnfoldState Fuse #-}
+data IterateUnfoldState o i =
+      IterateUnfoldOuter o
+    | IterateUnfoldInner o i [i]
+
+{-# INLINE_NORMAL iterateUnfoldManyDFS #-}
+iterateUnfoldManyDFS :: Monad m =>
+       Unfold m a a
+    -> Stream m a
+    -> Stream m a
+iterateUnfoldManyDFS (Unfold istep inject) (Stream ostep ost) =
+    Stream step (IterateUnfoldOuter ost)
+
+    where
+
+    {-# INLINE_LATE step #-}
+    step gst (IterateUnfoldOuter o) = do
```

Comment:
I wonder if you can match on only 1 constructor here.
You can give the same importance to the outer and inner states.
`IterateState currentState otherStates`

Reply:
Why have 2 different type variables? The `inject` is from the same unfold right?